dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,574 @@
+        for (auto & res : block.output_columns)
+        {
+            columns.push_back(ColumnPtr(std::move(res)));
+        }
+        ++next_output_block_number;
+
+        auto output_block = output_header.cloneWithColumns(std::move(columns));
+        releaseAlreadyOutputWindowBlock();
+        return output_block;
+    }
+    return {};
+}
+
+bool WindowBlockInputStream::onlyHaveRowNumber()
+{
+    return workspaces.size() == 1 && workspaces[0].window_function->getName() == "row_number";
+}
+
+bool WindowBlockInputStream::onlyHaveRowNumberAndRank()
+{
+    for (const auto & workspace : workspaces)
+    {
+        if (workspace.window_function->getName() != "row_number" && workspace.window_function->getName() != "rank" && workspace.window_function->getName() != "dense_rank")
+            return false;
+    }
+    return true;
+}
+
+void WindowBlockInputStream::releaseAlreadyOutputWindowBlock()
+{
+    // We don't really have to keep the entire partition, and it can be big, so
+    // we want to drop the starting blocks to save memory. We can drop the old
+    // blocks if we already returned them as output, and the frame and the
+    // current row are already past them. We also need to keep the previous
+    // frame start because we use it as the partition standard. It is always less
+    // than the current frame start, so we don't have to check the latter. Note
+    // that the frame start can be further than current row for some frame specs
+    // (e.g. EXCLUDE CURRENT ROW), so we have to check both.
+    assert(prev_frame_start <= frame_start);
+    const auto first_used_block = std::min(next_output_block_number,
+                                           std::min(prev_frame_start.block, current_row.block));
+
+    if (first_block_number < first_used_block)
+    {
+        window_blocks.erase(window_blocks.begin(),
+                            window_blocks.begin() + (first_used_block - first_block_number));
+        first_block_number = first_used_block;
+
+        assert(next_output_block_number >= first_block_number);
+        assert(frame_start.block >= first_block_number);
+        assert(prev_frame_start.block >= first_block_number);
+        assert(current_row.block >= first_block_number);
+    }
+}
+
+void WindowBlockInputStream::appendBlock(Block & current_block)
+{
+    if (!input_is_finished)
+    {
+        if (current_block.rows() == 0)
+        {
+            return;
+        }
+
+        window_blocks.push_back({});
+        auto & window_block = window_blocks.back();
+
+        window_block.rows = current_block.rows();
+
+        // Initialize output columns and add new columns to output block.
+        for (auto & ws : workspaces)
+        {
+            window_block.output_columns.push_back(ws.window_function->getReturnType()->createColumn());
+        }
+
+        window_block.input_columns = current_block.getColumns();
+    }
+
+    // Start the calculations. First, advance the partition end.
+    for (;;)
```

Comment:
Is it possible to separate the logic of the window function calculation into a new function?
like
```
appendBlock(current_block);
tryCalculate(); // other names are ok
```

Reply:
> Is it possible to separate the logic of the window function calculation into a new function? like
> 
> ```
> appendBlock(current_block);
> tryCalculate(); // other names are ok
> ```

done.